### PR TITLE
[enhancement](load) optimize load string data and dict page write

### PR DESCRIPTION
**Pull request by yangzhg**
# Proposed changes

1.  reduce condition check when memtable insert
2. reduce virtual function call of BinaryDictPageBuilder
3. remove `_dict_items`  add to `dict_builder` directly

## Problem Summary:

Describe the overview of changes.

## Checklist(Required)

1. Does it affect the original behavior: (Yes/No/I Don't know)
4. Has unit tests been added: (Yes/No/No Need)
5. Has document been added or modified: (Yes/No/No Need)
6. Does it need to update dependencies: (Yes/No)
7. Are there any changes that cannot be rolled back: (Yes/No)

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: be/src/olap/memtable.cpp [hunk 1]
```diff
@@ -65,44 +75,46 @@ int MemTable::RowCursorComparator::operator()(const char* left, const char* righ
     return compare_row(lhs_row, rhs_row);
 }
 
-void MemTable::insert(const Tuple* tuple) {
+// For non-DUP models, for the data rows passed from the upper layer, when copying the data,
+// we first allocate from _buffer_mem_pool, and then check whether it already exists in
+// _skiplist.  If it exists, we aggregate the new row into the row in skiplist.
+// otherwise, we need to copy it into _table_mem_pool before we can insert it.
+void MemTable::_insert_agg(const Tuple* tuple) {
```
Comment (zbtzbtzbt) [inline be/src/olap/memtable.cpp:82]: we may change a name, such as `_insert_no_dup`? and put  `_insert_dup` before `_insert_no_dup`.
becasue this kind of `insert` agg and unique data.

File: be/src/olap/rowset/segment_v2/bitshuffle_page.h [hunk 2]
```diff
@@ -93,13 +93,52 @@ class BitshufflePageBuilder : public PageBuilder {
...
         _count += to_add;
         _remain_element_capacity -= to_add;
         // return added number through count
         *count = to_add;
+        if constexpr (single) {
+            if constexpr (SIZE_OF_TYPE == 1) {
+                *reinterpret_cast<uint8_t*>(&_data[orig_size]) = *vals;
+                return Status::OK();
+            }
+            if constexpr (SIZE_OF_TYPE == 2) {
```
Comment (zbtzbtzbt) [inline be/src/olap/rowset/segment_v2/bitshuffle_page.h:123]: else if constexpr

File: be/src/olap/rowset/segment_v2/binary_plain_page.h [hunk 3]
```diff
@@ -127,11 +125,16 @@ class BinaryPlainPageBuilder : public PageBuilder {
         return Status::OK();
     }
 
-    void update_prepared_size(size_t added_size) {
-        _prepared_size += added_size;
-        _prepared_size += sizeof(uint32_t);
+    inline Slice operator[](std::size_t idx) const {
```
Comment (zbtzbtzbt) [inline be/src/olap/rowset/segment_v2/binary_plain_page.h:128]: size_t

File: be/src/olap/rowset/segment_v2/bitshuffle_page.h [hunk 4]
```diff
@@ -93,13 +93,52 @@ class BitshufflePageBuilder : public PageBuilder {
     bool is_page_full() override { return _remain_element_capacity == 0; }
 
     Status add(const uint8_t* vals, size_t* count) override {
+        return add_internal<false>(vals, count);
+    }
+
+    Status single_add(const uint8_t* vals, size_t* count) {
```
Comment (zbtzbtzbt) [inline be/src/olap/rowset/segment_v2/bitshuffle_page.h:99]: can we delete this function？

File: be/src/olap/rowset/segment_v2/binary_dict_page.cpp [hunk 5]
```diff
@@ -90,13 +92,18 @@ Status BinaryDictPageBuilder::add(const uint8_t* vals, size_t* count) {
...
+                    break;
+                }
                 _dictionary.emplace(dict_item, value_code);
-                _dict_items.push_back(dict_item);
-                _dict_builder->update_prepared_size(dict_item.size);
             }
             size_t add_count = 1;
-            RETURN_IF_ERROR(_data_page_builder->add(reinterpret_cast<const uint8_t*>(&value_code),
-                                                    &add_count));
+            RETURN_IF_ERROR(actual_builder->single_add(
```
Comment (zbtzbtzbt) [inline be/src/olap/rowset/segment_v2/binary_dict_page.cpp:105]: use `add_internal<true>`

File: be/src/olap/rowset/segment_v2/bitshuffle_page.h [hunk 6]
```diff
@@ -93,13 +93,52 @@ class BitshufflePageBuilder : public PageBuilder {
...
+                *reinterpret_cast<uint32_t*>(&_data[orig_size]) =
+                        *reinterpret_cast<const uint32_t*>(vals);
+                return Status::OK();
+            }
+            if constexpr (SIZE_OF_TYPE == 8) {
+                *reinterpret_cast<uint64_t*>(&_data[orig_size]) =
+                        *reinterpret_cast<const uint64_t*>(vals);
+                return Status::OK();
+            }
+            memcpy(&_data[orig_size], vals, to_add_size);
```
Comment (zbtzbtzbt) [inline be/src/olap/rowset/segment_v2/bitshuffle_page.h:138]: else

File: be/src/olap/rowset/segment_v2/bitshuffle_page.h [hunk 7]
```diff
@@ -93,13 +93,52 @@ class BitshufflePageBuilder : public PageBuilder {
     bool is_page_full() override { return _remain_element_capacity == 0; }
 
     Status add(const uint8_t* vals, size_t* count) override {
+        return add_internal<false>(vals, count);
```
Comment (zbtzbtzbt) [inline be/src/olap/rowset/segment_v2/bitshuffle_page.h:96]: can we delete this function?

**Review by zbtzbtzbt — COMMENTED** [review]
Thank you for your pr, I have learned a lot

File: be/src/olap/rowset/segment_v2/binary_plain_page.h [hunk 8]
```diff
@@ -47,14 +47,13 @@ namespace segment_v2 {
...
-            : _size_estimate(0), _prepared_size(0), _options(options) {
+            : _size_estimate(0), _options(options) {
         reset();
     }
 
     bool is_page_full() override {
         // data_page_size is 0, do not limit the page size
-        return _options.data_page_size != 0 && (_size_estimate > _options.data_page_size ||
-                                                _prepared_size > _options.data_page_size);
+        return _options.data_page_size != 0 && (_size_estimate > _options.data_page_size);
```
Comment (zbtzbtzbt) [inline be/src/olap/rowset/segment_v2/binary_plain_page.h:56]: better
`return (_options.data_page_size != 0 && _size_estimate > _options.data_page_size);`

Comment (yangzhg) [reply be/src/olap/rowset/segment_v2/binary_plain_page.h:56]: why ？

File: be/src/olap/memtable.cpp (hunk 1 above)
Comment (yangzhg) [reply be/src/olap/memtable.cpp:82]: The unique table is essentially an Agg table where the aggregation function on the value column is replace, and the use of unique is only for the convenience of optimization.

File: be/src/olap/rowset/segment_v2/binary_dict_page.cpp (hunk 5 above)
Comment (yangzhg) [reply be/src/olap/rowset/segment_v2/binary_dict_page.cpp:105]: The reason for adding add internal is that I want to unify the logic of add and single add, and I don't want to add a separate function, so add internal doesn't want to be exposed for use

File: be/src/olap/rowset/segment_v2/bitshuffle_page.h (hunk 4 above)
Comment (yangzhg) [reply be/src/olap/rowset/segment_v2/bitshuffle_page.h:99]: No，The reason for adding add internal is that I want to unify the logic of add and single add, and I don't want to add a separate function, so add internal doesn't want to be exposed for use

File: be/src/olap/rowset/segment_v2/bitshuffle_page.h (hunk 2 above)
Comment (yangzhg) [reply be/src/olap/rowset/segment_v2/bitshuffle_page.h:123]: else is no need this is like switch case  every if has a return and，and there is no overlap between if statements

File: be/src/olap/memtable.cpp (hunk 1 above)
Comment (zbtzbtzbt) [reply be/src/olap/memtable.cpp:82]: ok

File: be/src/olap/rowset/segment_v2/binary_plain_page.h (hunk 8 above)
Comment (zbtzbtzbt) [reply be/src/olap/rowset/segment_v2/binary_plain_page.h:56]: we might remove the `( )`
`return _options.data_page_size != 0 && _size_estimate > _options.data_page_size;`

or
`return (_options.data_page_size != 0) && (_size_estimate > _options.data_page_size);`

File: be/src/olap/rowset/segment_v2/bitshuffle_page.h (hunk 7 above)
Comment (yangzhg) [reply be/src/olap/rowset/segment_v2/bitshuffle_page.h:96]: no add is pure virtual

File: be/src/olap/rowset/segment_v2/bitshuffle_page.h [hunk 9]
```diff
@@ -93,13 +93,48 @@ class BitshufflePageBuilder : public PageBuilder {
...
+    }
+
+    Status single_add(const uint8_t* vals, size_t* count) {
+        return add_internal<true>(vals, count);
+    }
+
+    template <bool single>
+    inline Status add_internal(const uint8_t* vals, size_t* count) {
```
Comment (morningman) [inline be/src/olap/rowset/segment_v2/bitshuffle_page.h:104]: Why this implement faster than `_data.append`?

Comment (yangzhg) [reply be/src/olap/rowset/segment_v2/bitshuffle_page.h:104]: use assign instead of memcpy

**Review by morningman — DISMISSED** [review]
LGTM

**Review by morningman — APPROVED** [review]
LGTM

**Pull request merged**